File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_FRIENDSPOSITION.cs
using Atlasd.Battlenet.Exceptions;
using Atlasd.Daemon;
using System.IO;

namespace Atlasd.Battlenet.Protocols.Game.Messages
{
    class SID_FRIENDSPOSITION : Message
    {

        public SID_FRIENDSPOSITION()
        {
            Id = (byte)MessageIds.SID_FRIENDSPOSITION;
            Buffer = new byte[0];
        }

        public SID_FRIENDSPOSITION(byte[] buffer)
        {
            Id = (byte)MessageIds.SID_FRIENDSPOSITION;
            Buffer = buffer;
        }

        public override bool Invoke(MessageContext context)
        {
            if (context.Direction != MessageDirection.ServerToClient)
                throw new GameProtocolViolationException(context.Client, $"{MessageName(Id)} must be sent from server to client");

            /**
             * (UINT8) Old entry number
             * (UINT8) New entry number
             */

            Buffer = new byte[2];

            using var m = new MemoryStream(Buffer);
            using var w = new BinaryWriter(m);

            w.Write((byte)context.Arguments["old"]);
            w.Write((byte)context.Arguments["new"]);

            Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)");
            context.Client.Send(ToByteArray(context.Client.ProtocolType));
            return true;
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/Messages/SID_FRIENDSPOSITION.vb
Imports AtlasV.Battlenet.Exceptions
Imports AtlasV.Daemon
Imports System.IO

Namespace AtlasV.Battlenet.Protocols.Game.Messages
    Class SID_FRIENDSPOSITION
        Inherits Message

        Public Sub New()
            Id = CByte(MessageIds.SID_FRIENDSPOSITION)
            Buffer = Array.Empty(Of Byte)()
        End Sub

        Public Sub New(ByVal varBuffer As Byte())
            Id = CByte(MessageIds.SID_FRIENDSPOSITION)
            Buffer = varBuffer
        End Sub

        Public Overrides Function Invoke(ByVal context As MessageContext) As Boolean
            If context.Direction <> MessageDirection.ServerToClient Then Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} must be sent from server to client")
            Buffer = New Byte(1) {}

            Using m = New MemoryStream(Buffer)
                Using w = New BinaryWriter(m)
                    w.Write(CByte(context.Arguments("old")))
                    w.Write(CByte(context.Arguments("new")))
                End Using
            End Using

            Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
            context.Client.Send(ToByteArray(context.Client.ProtocolType))
            Return True
        End Function
    End Class
End Namespace
